Add managed packages & remote generation documentation (#220)
bufbuild/docs.buf.build: docs/bsr/remote-packages/migrating-from-alpha.md

---
id: migrating-from-alpha
title: Migrating from alpha
---

:::warning Deprecated

> Alpha templates will continue to work for a period of time before they are no longer supported.

Reviewer: Should have more details on when support will be removed for templates.

Reviewer 2: @bufdev how long will this be supported? would you like me to say "until early 2023"?

user.

Updated documentation can be found here:

- [Managed Packages](overview)
- [Usage](usage)

Existing remote generation [alpha documentation](../../reference/deprecated/remote-generation/overview.mdx) has
moved to the Reference section.

## Disable plugin uploads

You will no longer be able to upload plugins or create templates based on those
plugins.

For now, the Buf team will package and distribute plugins to the BSR, browsable
at [buf.build/plugins](https://buf.build/plugins). To learn more about how
plugins are packaged and distributed check out the [bufbuild/plugins
repository][bufbuild-plugins].

If you find a useful plugin that should be added please [file an
issue][bufbuild-plugins-issue]!

:::info What happens to old plugins and templates?

Existing plugins and templates on the [public BSR](https://buf.build) will
remain available, but will be removed at a later date. We take breaking changes
very seriously and want to provide ample opportunity for our valued users to
migrate and request plugins which have not yet been added.

User uploaded plugins will be supported for Enterprise customers.

:::

## Templates removed

> Templates were yet another concept for users to understand when interacting with

Reviewer: ```suggestion
Templates were a complex concept for users to understand when interacting with
```

> the BSR, so we removed them. Instead, you reference plugins directly by name. A

Reviewer: ```suggestion
the BSR, so we removed them. Instead, you reference plugins directly by name (and plugins can depend on the output of other plugins). A
```

list of plugins can be found at:

https://buf.build/plugins

All remote plugins can be used as `plugin` references in `buf.gen.yaml` files.

> Some plugins can be used with the BSR go module proxy and BSR npm registry.

Reviewer: ```suggestion
Some plugins can be used with the BSR Go module proxy and NPM registry.
```

### Versioning

The synthetic versioning scheme has been replaced by a more explicit versioning
scheme comprised of the plugin version, module reference (datetime+short commit
name) and revision. Example:

```
0.4.0-20220908151351-622fe7149695.1
```

This new semver-compatible versioning scheme can be pinned in lock files and
always references a specific plugin + module for reproducibility.

> Most users will be fetching `@latest` and will be unaffected by the versioning

Reviewer: ```suggestion
Most users fetch the `@latest` version and will be unaffected by the versioning
```

change.

## buf.gen.yaml

The buf.gen.yaml configuration file is largely unchanged, except for:

1. `name` and `remote` keys to reference plugins changed to just `plugin`. The
   `plugin` key understands both local and remote references. Requires [buf CLI
   version 1.8][buf-tag-18] or later.

1. Drop the `../plugins/..` path when referencing remote plugins

Full example covering both changes:

```diff
plugins:
-  - remote: buf.build/bufbuild/plugins/connect-go
+  - plugin: buf.build/bufbuild/connect-go
```

## Go module proxy

There are a couple of key changes from the alpha:

> - The base URL has changed to `buf.build/gen/go`

Reviewer: ```suggestion
- The base URL has changed from `go.buf.build` to `buf.build/gen/go`
```

- The path has changed to begin with the module name.
- The template reference in the path has been replaced with plugins and moved to
  the end.

> - The version has changed to inlcude plugin version and module commit

Reviewer: ```suggestion
- The version has changed to include plugin version and module commit
```

+     │   │   └── connect
+     │   │       └── demo
+     │   │           └── eliza
+     │   │               └── v1
+     │   │                   ├── eliza_connectweb.d.ts
+     │   │                   └── eliza_connectweb.js
+     │   └── package.json
+     └── bufbuild_eliza.bufbuild_es
+         ├── buf
+         │   └── connect
+         │       └── demo
+         │           └── eliza
+         │               └── v1
+         │                   ├── eliza_pb.d.ts
+         │                   └── eliza_pb.js
+         └── package.json
```

  </div>
</details>

Using this example, if your application code imported `eliza_pb.js` from
`@buf/bufbuild_connect-web_bufbuild_eliza/eliza_connectweb.js` then you'll want
to update that import within your application code to reference the base types
from `@buf/bufbuild_eliza.bufbuild_es/eliza_connectweb.js`. Assuming you
have `npm uninstall` and `npm install` based on the naming change mentioned
above.

### protocolbuffers/js and grpc/web templates

If you were using the [`protocolbuffers_js`][template-protocolbuffers-js] or
[`grpc_web`][template-grpc-web] templates there is no equivalent in the new BSR
npm registry.

> The plugins themselves are still hosted on the BSR and can be used with

Reviewer: Currently these have been removed (pending support for NPM registry). Do we want to re-add them or remove this from the docs for now?

`buf generate` commands.

> [js]: https://buf.build/bufbuild/es

Reviewer: Appears unused?

> [grpc-web]: https://buf.build/library/grpc-web

Reviewer: Also appears unused (and points to old location).
